### PR TITLE
[BCI-2714]: adds sncast scripts

### PR DESCRIPTION
**Pull request by chris-de-leon-cll**
# Adds `sncast` Scripts

## Overview

Please see the ticket [here](https://smartcontract-it.atlassian.net/browse/BCI-2714) for more details! tl;dr this PR adds support for using `sncast` to interact with the data feed example contracts on Starknet testnet OR a local container running [starknet-devnet-rs](https://github.com/0xSpaceShard/starknet-devnet-rs).

## Primary Changes

- [x] bumps starknet foundry from v0.18.0 to v0.21.0
- [x] bumps cairo to v2.6.3 
- [x] modifies `./ops/scripts/devnet-hardhat.sh` such that the container version can be configured (snfoundry v0.21.0 needs an rpc v0.7 endpoint, and the script defaults to using rpc v0.6)
- [x] adds the following data feed sncast scripts (all located in `./examples/contracts/aggregator_consumer/scripts/src/test-consumer`:
	- [x] deploy_test_consumer (deploys `MockAggregator` and `AggregatorConsumer`)
	- [x] read_decimals
	- [x] read_latest_round
	- [x] set_latest_round
- [x] adds helper commands in the Makefile to simplify script setup and execution
- [x] updates the `README.md` file in `./examples/contracts/aggregator_consumer`
	- [x] adds details on how to run the sncast scripts on both testnet and a local starknet devnet container using the helper commands in the Makefile
		- [x] describes what commands exist
		- [x] explains how to use the commands
		- [x] explains what each commands does
		- [x] explains how to read the output of the commands
- [x] simplifies the `consumer.cairo` contract
	- [x] removes `read_decimals` (not needed for new docs)
	- [x] removes `read_latest_round` (not needed for new docs)
	- [x] adds `read_answer` (needed for new docs)
	- [x] adds `set_answer` (needed for new docs)
	- [x] updates the test cases in `test_consumer.cairo` accordingly

### CODE REVIEW

File: examples/contracts/aggregator_consumer/Makefile [hunk 1]
```diff
@@ -1,12 +1,53 @@
...
+# supports rpc v0.7 which is needed for snfoundry v0.20.1
+CONTAINER_VERSION="7743a089a33beb75d7012e4aa24745bee8ae0d71"
+
+# The name of the account that will be used for snfoundry
+# scripts + transactions
+ACCOUNT_NAME="test-acct"
+
+devnet:
+	@printf "\nStarting a local starknet devnet docker container:\n\n" \
+		&& bash ../../../ops/scripts/devnet-hardhat-down.sh \
```
Comment (archseer) [inline examples/contracts/aggregator_consumer/Makefile]: We already call cleanup on startup: https://github.com/smartcontractkit/chainlink-starknet/blob/38b10b2e272aed56d0538f39aee4b2e89411753a/ops/scripts/devnet-hardhat.sh#L11

File: examples/contracts/aggregator_consumer/Makefile [hunk 2]
```diff
@@ -1,12 +1,53 @@
...
+	@sncast account deploy --name $(ACCOUNT_NAME)
+
+delete-account:
+	@sncast account delete --name $(ACCOUNT_NAME)
+
+tc-deploy:
+	@cd ./scripts && sncast --account "$(ACCOUNT_NAME)" script run deploy_test_consumer
+
+tc-set-latest-round:
+	@cd ./scripts && sncast --account "$(ACCOUNT_NAME)" script run set_latest_round
```
Comment (archseer) [inline examples/contracts/aggregator_consumer/Makefile]: You can define `account = "$ACCOUNT_NAME"` in the toml config under the default profile, then we don't need to specify `--account` everywhere

File: examples/contracts/aggregator_consumer/README.md [hunk 3]
```diff
@@ -2,26 +2,179 @@
...
+  url = "https://starknet-sepolia.public.blastapi.io/rpc/v0_7"
+  ```
+
+- Next, we'll need to create an account. This account will be used to pay for testnet transaction fees for some of the scripts. To generate an account (but not necessarily deploy it to the network yet), you can run the following command:
+
+  ```sh
+  make create-account
+  ```
+  
+- Once the account has been created, its info should be stored in an accounts file on your local machine (usually at `~/.starknet_accounts/starknet_open_zeppelin_accounts.json`). If you'd like to view the account information later on, you can use the following command:
```
Comment (archseer) [inline examples/contracts/aggregator_consumer/README.md]: Let's define `accounts-file` to be in the current directory and gitignore it. That way we're not leaving global artifacts on user's system

Comment (archseer) [reply examples/contracts/aggregator_consumer/README.md]: (only for devnet though)

File: examples/contracts/aggregator_consumer/Makefile [hunk 4]
```diff
@@ -1,12 +1,53 @@
...
+	@printf "\nCreating an account named \"$(ACCOUNT_NAME)\"...\n\n" \
+		&& sncast account create --name $(ACCOUNT_NAME) \
+	  && printf "\nYour accounts:\n\n" \
+		&& cat ~/.starknet_accounts/starknet_open_zeppelin_accounts.json \
+		&& printf "\n\nYou can fund your account here: https://sepolia.starkgate.starknet.io\n"
+
+add-account:
+	@printf "\nAdding an account named \"$(ACCOUNT_NAME)\"...\n\n" \
+		&& sncast \
+			--url "http://127.0.0.1:5050/rpc" \
```
Comment (archseer) [inline examples/contracts/aggregator_consumer/Makefile]: How about a devnet and a testnet profile in sncast, then removing `--url` flags and hardcoding the account name on devnet?

File: examples/contracts/aggregator_consumer/Makefile (hunk 1 above)
Comment (chris-de-leon-cll) [reply examples/contracts/aggregator_consumer/Makefile]: Good catch! I'll modify the script

File: examples/contracts/aggregator_consumer/README.md (hunk 3 above)
Comment (chris-de-leon-cll) [reply examples/contracts/aggregator_consumer/README.md]: Done!

- In `snfoundry.toml`, I made `accounts-file` configurable through an environment variable which is set automatically in the `Makefile`
- The `Makefile` is now the only place where the `accounts-file` paths for devnet and testnet are defined - it is not hardcoded in any other place
- For devnet, the accounts file is called `accounts.json` and it lives at the same level as `snfoundry.toml`. It is currently gitignored using a pre-existing rule in the `.gitignore` file in the repo's root directory
- For testnet, the accounts file is unchanged (`~/.starknet_accounts/starknet_open_zeppelin_accounts.json`)

File: examples/contracts/aggregator_consumer/Makefile (hunk 2 above)
Comment (chris-de-leon-cll) [reply examples/contracts/aggregator_consumer/Makefile]: Done!

- In `snfoundry.toml` the account name is now configurable through an environment variable which is set automatically in the `Makefile`
- The `Makefile` is now the only place where the account names for devnet and testnet are defined - it is not hardcoded in any other place

File: examples/contracts/aggregator_consumer/Makefile (hunk 4 above)
Comment (chris-de-leon-cll) [reply examples/contracts/aggregator_consumer/Makefile]: Done!

- In `snfoundry.toml`, there is now a devnet profile and testnet profile each with their own RPC URLs
- In the `Makefile` the commands now use the `--profile` flag
- The `Makefile` sncast commands now require an argument `NETWORK` which is used to determine which profile a.k.a which network to run the script on

File: examples/contracts/aggregator_consumer/Makefile [hunk 5]
```diff
@@ -1,12 +1,73 @@
...
+	@printf "\nImporting a prefunded account from starknet devnet container...\n\n" \
+		&& sncast --profile devnet account add \
+			--name "$(DEVNET_ACCOUNT_NAME)" \
+			--address "0x4b3f4ba8c00a02b66142a4b1dd41a4dfab4f92650922a3280977b0f03c75ee1" \
+			--private-key "0x57b2f8431c772e647712ae93cc616638" \
+	  && printf "\nYour accounts:\n\n" \
+		&& cat $(DEVNET_ACCOUNTS_FILE) \
+
+deploy-account:
+	@sncast --profile testnet account deploy
```
Comment (khadni) [inline examples/contracts/aggregator_consumer/Makefile]:
```suggestion
	@sncast --profile testnet account deploy --name "$(TESTNET_ACCOUNT_NAME)"
```

File: examples/contracts/aggregator_consumer/README.md [hunk 6]
```diff
@@ -2,26 +2,270 @@
...
+- `src/`: contains sample cairo contracts that demonstrate how one can integrate with Chainlink's core starknet contracts.
+- `tests/`: contains cairo tests for the example contracts in `src/`. They showcase some simple usage patterns for the contracts.
+- `scripts/`: contains cairo scripts that allow you to interact with the example contracts over testnet or a local starknet devnet container. 
 
+## Prerequisites
 
 To get started, ensure that you have the following tools installed on your machine:
 
-- [starknet-foundry (v0.18.0)](https://github.com/foundry-rs/starknet-foundry/releases/tag/v0.18.0)
+- [starknet-foundry (v0.20.1)](https://github.com/foundry-rs/starknet-foundry/releases/tag/v0.18.0)
```
Comment (khadni) [inline examples/contracts/aggregator_consumer/README.md]:
```suggestion
- [starknet-foundry (v0.20.1)](https://github.com/foundry-rs/starknet-foundry/releases/tag/v0.20.1)
```

Comment (chris-de-leon-cll) [reply examples/contracts/aggregator_consumer/README.md]: Thank you! Updated the link 👍

File: examples/contracts/aggregator_consumer/Makefile (hunk 5 above)
Comment (chris-de-leon-cll) [reply examples/contracts/aggregator_consumer/Makefile]: Good catch - added the additional flag!

**Review by archseer — APPROVED** [review]

**Pull request merged**